File: tools/typescript/src/session.ts
// AgentOS TypeScript SDK Session
// Version: 1.0.0.5
// Last updated: 2026-03-21

import { SessionError } from './errors';
import { AgentOS } from './agent';

/**
 * Session class for managing AgentOS sessions
 */
export class Session {
  private client: AgentOS;
  private sessionId: string;

  /**
   * Create a new Session object
   * @param client AgentOS client
   * @param sessionId Session ID
   */
  constructor(client: AgentOS, sessionId: string) {
    this.client = client;
    this.sessionId = sessionId;
  }

  /**
   * Get the session ID
   * @returns Session ID
   */
  get id(): string {
    return this.sessionId;
  }

  /**
   * Set a context value for the session
   * @param key Context key
   * @param value Context value
   * @returns True if the context was set successfully
   */
  async setContext(key: string, value: any): Promise<boolean> {
    try {
      const response = await this.client['request']<{ success: boolean }>(
        'POST',
        `/api/sessions/${this.sessionId}/context`,
        { key, value }
      );

      return response.success;
    } catch (error) {
      throw new SessionError(`Error setting session context: ${error.message}`);
    }
  }

  /**
   * Get a context value from the session
   * @param key Context key
   * @returns Context value
   */
  async getContext(key: string): Promise<any> {
    try {
      const response = await this.client['request']<{ value: any }>(
        'GET',
        `/api/sessions/${this.sessionId}/context/${key}`
      );

      return response.value;
    } catch (error) {
      throw new SessionError(`Error getting session context: ${error.message}`);
    }
  }

  /**
   * Close the session
   * @returns True if the session was closed successfully
   */
  async close(): Promise<boolean> {
    try {
      const response = await this.client['request']<{ success: boolean }>(
        'DELETE',
        `/api/sessions/${this.sessionId}`
      );

      return response.success;
    } catch (error) {
      throw new SessionError(`Error closing session: ${error.message}`);
    }
  }
}

File: tools/rust/src/session.rs
// AgentOS Rust SDK Session
// Version: 1.0.0.5
// Last updated: 2026-03-21

use serde_json::{json, Value};

use crate::{AgentOSError, Client};

/// Session
type Result<T> = std::result::Result<T, AgentOSError>;

/// Session
type JsonObject = serde_json::Map<String, Value>;

/// Session
#[derive(Debug, Clone)]
pub struct Session {
    client: Client,
    id: String,
}

impl Session {
    /// Create a new Session object
    pub fn new(client: Client, id: String) -> Self {
        Session {
            client,
            id,
        }
    }
    
    /// Set a context value for the session
    pub async fn set_context(&self, key: &str, value: Value) -> Result<bool> {
        let path = format!("/api/sessions/{}/context", self.id);
        let data = json!({"key": key, "value": value});
        let response = self.client.request("POST", &path, Some(&data)).await?;
        
        let success = response.get("success")
            .and_then(|v| v.as_bool())
            .ok_or_else(|| AgentOSError::InvalidResponse("Missing success".to_string()))?;
        
        Ok(success)
    }
    
    /// Get a context value from the session
    pub async fn get_context(&self, key: &str) -> Result<Value> {
        let path = format!("/api/sessions/{}/context/{}", self.id, key);
        let response = self.client.request("GET", &path, None).await?;
        
        let value = response.get("value")
            .ok_or_else(|| AgentOSError::InvalidResponse("Missing value".to_string()))?;
        
        Ok(value.clone())
    }
    
    /// Close the session
    pub async fn close(&self) -> Result<bool> {
        let path = format!("/api/sessions/{}", self.id);
        let response = self.client.request("DELETE", &path, None).await?;
        
        let success = response.get("success")
            .and_then(|v| v.as_bool())
            .ok_or_else(|| AgentOSError::InvalidResponse("Missing success".to_string()))?;
        
        Ok(success)
    }
}
